core/templates/pages/teach-page/teach-page.controller.spec.ts

```diff
@@ -0,0 +1,146 @@
+// Copyright 2014 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for teachPage.
+ */
+import { WindowRef } from 'services/contextual/window-ref.service';
+
+describe('Teach Page', function() {
+  var $scope = null, ctrl = null;
+  var $timeout = null;
+  var SiteAnalyticsService = null;
+  var windowRef = new WindowRef();
+
+  beforeEach(angular.mock.module('oppia'));
+  beforeEach(angular.mock.module('oppia', function($provide) {
+    $provide.value('WindowRef', windowRef);
+  }));
+  beforeEach(angular.mock.inject(function($injector) {
+    $timeout = $injector.get('$timeout');
+    SiteAnalyticsService = $injector.get('SiteAnalyticsService');
+
+    var $rootScope = $injector.get('$rootScope');
+    $scope = $rootScope.$new();
+    var directive = $injector.get('teachPageDirective')[0];
+    ctrl = $injector.instantiate(directive.controller, {
+      $rootScope: $scope
+    });
+  }));
+
+  afterEach(function() {
+    // onhashchange and location.hash are reassigned because it shares
```

Comment:
Same comments for this, ts-ignore, etc. as in previous files.

Reply:
Done.